### PR TITLE
CON-3826 - Add Support for IA Policy

### PR DESCRIPTION
**Pull request by barnuri-cp**

### CODE REVIEW

File: dome9/common/resourcetype/resource_type.go [hunk 1]
```diff
@@ -24,4 +24,5 @@ const (
 	AwsUnifiedOnboarding                         = "dome9_aws_unified_onboarding"
 	AdmissionControlPolicy                       = "dome9_admission_control_policy"
 	Assessment                                   = "dome9_assessment"
+	ImageAssurancePolicy                         = "dome9_imageassurance_policy"
```
Comment (omartawfeeq) [inline dome9/common/resourcetype/resource_type.go]: can you change the name to dome9_image_assurance_policy to make it more readable ?

File: dome9/data_source_dome9_imageassurance_policy.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
+package dome9
```
Comment (omartawfeeq) [inline dome9/data_source_dome9_imageassurance_policy.go]: also here to change the file name to ...image_assurance_policy.go

File: dome9/data_source_dome9_imageassurance_policy.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+			},
+		},
+	}
+}
+
+func dataSourceImageAssurancePolicyRead(d *schema.ResourceData, meta interface{}) error {
+	d9Client := meta.(*Client)
+
+	policyID := d.Get("id").(string)
+	log.Printf("Getting data for Admission Control Policy id: %s\n", policyID)
```
Comment (omartawfeeq) [inline dome9/data_source_dome9_imageassurance_policy.go]: Change the log to Image Assurance Policy

File: dome9/provider.go [hunk 4]
```diff
@@ -80,6 +81,7 @@ func Provider() terraform.ResourceProvider {
 			resourcetype.ServiceAccount:                               dataSourceServiceAccount(),
 			resourcetype.AdmissionControlPolicy:                       dataSourceAdmissionControlPolicy(),
 			resourcetype.Assessment:                                   dataSourceAssessment(),
+			resourcetype.ImageAssurancePolicy:                         dataSourceAdmissionControlPolicy(),
```
Comment (omartawfeeq) [inline dome9/provider.go]: change to  dataSourceImageAssurancePolicyRead()

File: examples/imageassurance_policy/main.tf [hunk 5]
```diff
@@ -0,0 +1,31 @@
+# Configure the Dome9 Provider, works with US region by default
+provider "dome9" {
+  dome9_access_id     = "DOME9_API_ACCESS_ID"
+  dome9_secret_key    = "DOME9_API_SECRET"
+}
+
+resource "dome9_admission_control_policy" "admission_control_policy_test" {
```
Comment (omartawfeeq) [inline examples/imageassurance_policy/main.tf]: change the resource type and name to be suitable with Image Assurance Policy

File: examples/imageassurance_policy/main.tf [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  target_id = "1212c5a5-ccc7-451a-8b3f-f19b5fc041f4"
+  ruleset_id = -2001
+  // options: ["Environment", "OrganizationalUnit"]
+  target_type = "Environment"
+  notification_ids = ["52008dbc-6fdc-45d5-b588-de2afe13ac5d"]
+  admission_control_action = "Detection"
+  admission_control_unscanned_action = "Detection"
+}
+
+data "dome9_admission_control_policy" "admission_control_policy_test" {
```
Comment (omartawfeeq) [inline examples/imageassurance_policy/main.tf]: change the data type and name to be suitable with Image Assurance Policy

File: vendor/github.com/dome9/dome9-sdk-go/services/imageassurance/policy/imageassurance_policy.go [hunk 7]
```diff
@@ -1,35 +1,42 @@
-package admission_policy
+package imageassurance_policy
```
Comment (omartawfeeq) [inline vendor/github.com/dome9/dome9-sdk-go/services/imageassurance/policy/imageassurance_policy.go]: here is not sure why github detect a file change/modification instead of addition

File: website/docs/d/imageassurance_policy.html.markdown [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `target_id` - Environment ID / OU ID.
+* `target_type` - Policy Type (`Environment`, `OrganizationalUnit`).
+* `ruleset_id` - Ruleset ID.
+* `notification_ids` - Notification IDs [list].
+* `admission_control_action` - Policy action type (`Detection`, `Prevention`).
```
Comment (omartawfeeq) [inline website/docs/d/imageassurance_policy.html.markdown]: Can you change Policy to Image Admission Policy

File: website/docs/d/imageassurance_policy.html.markdown [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+* `admission_control_unscanned_action` - Policy action type (`Detection`, `Prevention`).
```
Comment (omartawfeeq) [inline website/docs/d/imageassurance_policy.html.markdown]: Can you change Policy to Image Admission Policy

File: website/docs/r/imageassurance_policy.html.markdown [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+layout: "dome9"
+page_title: "Check Point CloudGuard Dome9: dome9_imageassurance_policy"
+sidebar_current: "docs-resource-dome9-admission-control-policy"
+description: |-
+  Creates imageassurance policies in Dome9
+---
+
+# dome9_imageassurance_policy
+
+This resource is used to create and modify imageassurance policy in CloudGuard for Kubernetes. An imageassurance policy is the combination of a Ruleset applied to a specific Kubernetes environment with specific action, ruleset type have to be 'kubernetesruntimeassurance''.
```
Comment (omartawfeeq) [inline website/docs/r/imageassurance_policy.html.markdown]: remove ruleset type have to be 'kubernetesruntimeassurance''

File: website/docs/r/imageassurance_policy.html.markdown [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+## Example Usage
+
+Basic usage:
+
+```hcl
+resource "dome9_imageassurance_policy" "test_ac_policy" {
```
Comment (omartawfeeq) [inline website/docs/r/imageassurance_policy.html.markdown]: change test_ac_policy to test_ia_policy

File: website/docs/r/imageassurance_policy.html.markdown [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  target_id    = "Environment ID"
+  ruleset_id   = 00000
+  target_type  = "Environment"
+  notification_ids    = ["NOTIFICATION IDS"]
+  action       = "Detection"
```
Comment (omartawfeeq) [inline website/docs/r/imageassurance_policy.html.markdown]: here needs to use `admission_control_action` and `admission_control_unscanned_action`

**Comment by omartawfeeq** [comment]
Great Job Bar, needs couple of important things to do : 
1- Add Go tests for both the resource and data source : Example : https://github.com/dome9/terraform-provider-dome9/blob/master/dome9/data_source_dome9_admission_control_policy_test.go
2- As i know, you need to add all the provider sdk operations inside this project (dome9-sdk-go) : example : https://github.com/dome9/dome9-sdk-go/pull/114
3- as i remember, needs to run (go mod tidy && go mod vendor) to update go.sum file

File: dome9/common/resourcetype/resource_type.go (hunk 1 above)
Comment (barnuri-cp) [reply dome9/common/resourcetype/resource_type.go]: changed

File: dome9/data_source_dome9_imageassurance_policy.go (hunk 2 above)
Comment (barnuri-cp) [reply dome9/data_source_dome9_imageassurance_policy.go]: changed

File: dome9/data_source_dome9_imageassurance_policy.go (hunk 3 above)
Comment (barnuri-cp) [reply dome9/data_source_dome9_imageassurance_policy.go]: changed

File: dome9/provider.go (hunk 4 above)
Comment (barnuri-cp) [reply dome9/provider.go]: changed

File: examples/imageassurance_policy/main.tf (hunk 5 above)
Comment (barnuri-cp) [reply examples/imageassurance_policy/main.tf]: updated

File: examples/imageassurance_policy/main.tf (hunk 6 above)
Comment (barnuri-cp) [reply examples/imageassurance_policy/main.tf]: update

File: website/docs/d/imageassurance_policy.html.markdown (hunk 8 above)
Comment (barnuri-cp) [reply website/docs/d/imageassurance_policy.html.markdown]: changed

File: website/docs/d/imageassurance_policy.html.markdown (hunk 9 above)
Comment (barnuri-cp) [reply website/docs/d/imageassurance_policy.html.markdown]: changed

File: website/docs/r/imageassurance_policy.html.markdown (hunk 11 above)
Comment (barnuri-cp) [reply website/docs/r/imageassurance_policy.html.markdown]: changed

File: website/docs/r/imageassurance_policy.html.markdown (hunk 10 above)
Comment (barnuri-cp) [reply website/docs/r/imageassurance_policy.html.markdown]: changed

File: dome9/data_source_dome9_image_assurance_policy.go [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+			},
+		},
+	}
+}
+
+func dataSourceImageAssurancePolicyRead(d *schema.ResourceData, meta interface{}) error {
+	d9Client := meta.(*Client)
+
+	policyID := d.Get("id").(string)
+	log.Printf("Getting data for Admission Control Policy id: %s\n", policyID)
```
Comment (omartawfeeq) [inline dome9/data_source_dome9_image_assurance_policy.go]: can you change from "Getting data for Admission Control Policy" to "Getting data for Image Assurance Policy", don't know why i'm still seeing it, it is minor

File: website/docs/r/imageassurance_policy.html.markdown (hunk 12 above)
Comment (omartawfeeq) [reply website/docs/r/imageassurance_policy.html.markdown]: here still not changed since no 'action' in the image assurance resource

File: .gitignore [hunk 14]
```diff
@@ -1,2 +1,3 @@
 
 .idea/
+terraform-provider-dome9*
```
Comment (omartawfeeq) [inline .gitignore:3]: this means we are ignoring all the dirctory for terraform-provider-dome9 from git ?

File: .vscode/settings.json [hunk 15]
```diff
@@ -0,0 +1,3 @@
+{
+    "go.inferGopath": false
+}
```
Comment (omartawfeeq) [inline .vscode/settings.json]: dp we need these special settings related to special ide ? i think they are uncessary since related only to vscode

File: dome9/data_source_dome9_image_assurance_policy_test.go [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+		variable.CloudAccountKubernetesRuntimeProtectionEnabled,
+		variable.CloudAccountKubernetesAdmissionControlEnabled,
+		variable.CloudAccountKubernetesImageAssuranceEnabled,
+		variable.CloudAccountKubernetesThreatIntelligenceEnabled)
+
+	// Generate Notification HCL Configurations
+	notificationHCL := getContinuousComplianceNotificationResourceHCL(notificationGeneratedName, continuousComplianceNotificationConfig())
+
+	// Generate Image Assurance Policy HCL Resource
+	admissionPolicyHCL := getImageAssurancePolicyResourceHCL(kubernetesAccountHCL, kubernetesAccountResourceTypeAndName, notificationHCL,
```
Comment (omartawfeeq) [inline dome9/data_source_dome9_image_assurance_policy_test.go]: please change to imageAssurancePolicy...

File: dome9/data_source_dome9_image_assurance_policy_test.go [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+		PreCheck: func() {
+			testAccPreCheck(t)
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckImageAssurancePolicyDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckImageAssurancePolicyBasic(admissionPolicyHCL, policyGeneratedName, policyTypeAndName),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttrPair(policyDataSourceTypeAndName, "action", policyTypeAndName, "action"),
```
Comment (omartawfeeq) [inline dome9/data_source_dome9_image_assurance_policy_test.go]: maybe the test will fail here since action attribute is not inside Image Policy resource

File: dome9/resource_dome9_image_assurance_policy_test.go [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+// continuous compliance notification resource
+%s
+
+// admission control policy resource creation
+resource "%s" "%s" {
+  target_id    = "${%s.id}"
+  ruleset_id   = "%d"
+  target_type  = "%s"
+  notification_ids    = ["${%s.id}"]
+  action       = "%s"
```
Comment (omartawfeeq) [inline dome9/resource_dome9_image_assurance_policy_test.go]: also here needs to handle image admission action instead of action

File: website/docs/r/imageassurance_policy.html.markdown (hunk 12 above)
Comment (omartawfeeq) [reply website/docs/r/imageassurance_policy.html.markdown]: still i see it not changed here!

File: dome9/data_source_dome9_image_assurance_policy.go (hunk 13 above)
Comment (omartawfeeq) [reply dome9/data_source_dome9_image_assurance_policy.go]: also same here, need to change

File: dome9/resource_dome9_image_assurance_policy_test.go (hunk 18 above)
Comment (omartawfeeq) [reply dome9/resource_dome9_image_assurance_policy_test.go]: also here

File: .vscode/settings.json (hunk 15 above)
Comment (barnuri-cp) [reply .vscode/settings.json]: deleted

File: dome9/common/testing/variable/variable.go [hunk 19]
```diff
@@ -187,6 +187,14 @@ const (
 	AdmissionControlKubernetesAccountName  = "test_admission_control_kubernetes_resource_name"
 )
 
+const (
+	ImageAssurancePolicyDefaultRulesetId = -2001
```
Comment (omartawfeeq) [inline dome9/common/testing/variable/variable.go]: the ia default ruleset id is -2002 and not -2001

File: dome9/provider.go [hunk 20]
```diff
@@ -57,6 +57,7 @@ func Provider() terraform.ResourceProvider {
 			resourcetype.ServiceAccount:                   resourceServiceAccount(),
 			resourcetype.AdmissionControlPolicy:           resourceAdmissionPolicy(),
 			resourcetype.Assessment:                       resourceAssessment(),
+			resourcetype.ImageAssurancePolicy:             resourceAdmissionPolicy(),
```
Comment (omartawfeeq) [inline dome9/provider.go]: need to change to the required Image Assurance resource resourceImageAssurancePolicy()

File: examples/image_assurance_policy/main.tf [hunk 21]
```diff
@@ -0,0 +1,35 @@
+# Configure the Dome9 Provider, works with US region by default
+provider "dome9" {
+  dome9_access_id     = "DOME9_API_ACCESS_ID"
+  dome9_secret_key    = "DOME9_API_SECRET"
+}
+
+resource "dome9_image_assurance_policy" "image_assurance_policy_test" {
+  target_id = "1212c5a5-ccc7-451a-8b3f-f19b5fc041f4"
+  ruleset_id = -2001
```
Comment (omartawfeeq) [inline examples/image_assurance_policy/main.tf]: change to -2002

Comment (barnuri-cp) [reply examples/image_assurance_policy/main.tf]: fixed

File: dome9/provider.go (hunk 20 above)
Comment (barnuri-cp) [reply dome9/provider.go]: fixed

File: dome9/common/testing/variable/variable.go (hunk 19 above)
Comment (barnuri-cp) [reply dome9/common/testing/variable/variable.go]: fixed

File: dome9/resource_dome9_image_assurance_policy_test.go [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+		}
+
+		ImageAssurancePolicyResponse, _, err := apiClient.imageAssurancePolicy.Get(rs.Primary.ID)
+
+		if err == nil {
+			return fmt.Errorf("id %s already exists", rs.Primary.ID)
+		}
+
+		if ImageAssurancePolicyResponse != nil {
+			return fmt.Errorf("admission control policy with id %s exists and wasn't destroyed", rs.Primary.ID)
```
Comment (omartawfeeq) [inline dome9/resource_dome9_image_assurance_policy_test.go]: can you change to image assurance  ?

File: dome9/resource_dome9_image_assurance_policy_test.go [hunk 23]
```diff
@@ -0,0 +1,177 @@
...
+	kubernetesAccountHCL := getCloudAccountKubernetesResourceHCLWithfeatures(kubernetesAccountGeneratedName, variable.ImageAssuranceKubernetesAccountName,
+		variable.CloudAccountKubernetesRuntimeProtectionEnabled,
+		variable.CloudAccountKubernetesAdmissionControlEnabled,
+		variable.CloudAccountKubernetesImageAssuranceEnabled,
+		variable.CloudAccountKubernetesThreatIntelligenceEnabled)
+
+	// Generate Compliance Notification HCL Resource
+	notificationHCL := getContinuousComplianceNotificationResourceHCL(notificationGeneratedName, continuousComplianceNotificationConfig())
+
+	// Generate Admission Control Policy HCL Resource
```
Comment (omartawfeeq) [inline dome9/resource_dome9_image_assurance_policy_test.go]: also here if you can change the comment and variable names for ia

**Review by omartawfeeq — APPROVED** [review]
Good Job Bar

File: dome9/resource_dome9_image_assurance_policy.go [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+		},
+		Schema: map[string]*schema.Schema{
+			"target_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"target_type": {
+				Type:         schema.TypeString,
+				Required:     true,
```
Comment (ronyscp) [inline dome9/resource_dome9_image_assurance_policy.go:29]: isn't it optional?

File: dome9/resource_dome9_image_assurance_policy.go [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+	return resourceImageAssurancePolicyRead(d, meta)
+}
+
+func expandImageAssurancePolicyRequest(d *schema.ResourceData) imageassurance_policy.ImageAssurancePolicyRequest {
+	return imageassurance_policy.ImageAssurancePolicyRequest{
+		TargetId:                        d.Get("target_id").(string),
+		RulesetId:                       d.Get("ruleset_id").(int),
+		NotificationIds:                 expandNotificationIDs(d, "notification_ids"),
+		TargetType:                      d.Get("target_type").(string),
+		AdmissionControllerAction:       d.Get("admission_control_action").(string),
```
Comment (ronyscp) [inline dome9/resource_dome9_image_assurance_policy.go:76]: cast on potential nil?

File: website/docs/r/image_assurance_policy.html.markdown [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+-   `target_id` - (Required) The kubernetes environment id / organizational unit id.
+-   `ruleset_id` - (Required) The bundle id for the bundle that will be used in the policy.
+-   `target_type` - (Required) The imageassurance policy type ("Environment", "OrganizationalUnit").
+-   `notification_ids` - (Required) The notification policy id's for the policy [list].
+-   `admission_control_action` - (Required) The imageassurance policy action ("Prevention", "Detection").
```
Comment (ronyscp) [inline website/docs/r/image_assurance_policy.html.markdown]: optional (both actions)

File: dome9/resource_dome9_image_assurance_policy_test.go [hunk 27]
```diff
@@ -0,0 +1,177 @@
...
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckImageAssurancePolicyDestroy,
+		Steps: []resource.TestStep{
+			{
+				// Create Policy Test Step
+				Config: testAccCheckImageAssurancePolicyBasic(imageAssurancePolicyHCL, policyGeneratedName, policyTypeAndName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckImageAssurancePolicyExists(policyTypeAndName, &response),
+					resource.TestCheckResourceAttr(policyTypeAndName, "admission_control_action", variable.ImageAssurancePolicyDetectAction),
```
Comment (ronyscp) [inline dome9/resource_dome9_image_assurance_policy_test.go:53]: please add a test where the actions are nil

File: dome9/resource_dome9_image_assurance_policy_test.go [hunk 28]
```diff
@@ -0,0 +1,177 @@
...
+		}
+
+		ImageAssurancePolicyResponse, _, err := apiClient.imageAssurancePolicy.Get(rs.Primary.ID)
+
+		if err == nil {
+			return fmt.Errorf("id %s already exists", rs.Primary.ID)
+		}
+
+		if ImageAssurancePolicyResponse != nil {
+			return fmt.Errorf("imageAssurance policy with id %s exists and wasn't destroyed", rs.Primary.ID)
```
Comment (ronyscp) [inline dome9/resource_dome9_image_assurance_policy_test.go]: in all logs and comments please change imageAssurance to Image Assurance

File: website/docs/r/image_assurance_policy.html.markdown [hunk 29]
```diff
@@ -0,0 +1,52 @@
+---
+layout: 'dome9'
+page_title: 'Check Point CloudGuard Dome9: dome9_image_assurance_policy'
+sidebar_current: 'docs-resource-dome9-admission-control-policy'
```
Comment (ronyscp) [inline website/docs/r/image_assurance_policy.html.markdown]: *image assurance

File: dome9/resource_dome9_image_assurance_policy_test.go [hunk 30]
```diff
@@ -0,0 +1,177 @@
...
+					resource.TestCheckResourceAttr(policyTypeAndName, "ruleset_id", strconv.Itoa(variable.ImageAssurancePolicyDefaultRulesetId)),
+					resource.TestCheckResourceAttrSet(policyTypeAndName, "target_id"),
+					resource.TestCheckResourceAttr(policyTypeAndName, "target_type", variable.ImageAssurancePolicyTargetType),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckImageAssurancePolicyExists(resource string, acPolicy *imageAssurancePolicy.ImageAssurancePolicyResponse) resource.TestCheckFunc {
```
Comment (ronyscp) [inline dome9/resource_dome9_image_assurance_policy_test.go]: not acPolicy

File: dome9/resource_dome9_image_assurance_policy.go (hunk 24 above)
Comment (barnuri-cp) [reply dome9/resource_dome9_image_assurance_policy.go:29]: no, its required
https://github.com/dome9/dome9_backend/blob/master/API/ApiConSec/Dome9.ConSec.WebApi/Controllers/K8s/ImageAssurance/ImageAssurancePolicyRequestModel.cs#L13-L14

File: dome9/resource_dome9_image_assurance_policy.go (hunk 25 above)
Comment (barnuri-cp) [reply dome9/resource_dome9_image_assurance_policy.go:76]: what do you mean ? (not so familiar with go most of the code is copy paste from AC)

File: website/docs/r/image_assurance_policy.html.markdown (hunk 26 above)
Comment (barnuri-cp) [reply website/docs/r/image_assurance_policy.html.markdown]: fixed message

File: dome9/resource_dome9_image_assurance_policy_test.go (hunk 28 above)
Comment (barnuri-cp) [reply dome9/resource_dome9_image_assurance_policy_test.go]: fixed

File: website/docs/r/image_assurance_policy.html.markdown (hunk 29 above)
Comment (barnuri-cp) [reply website/docs/r/image_assurance_policy.html.markdown]: fixed

File: dome9/resource_dome9_image_assurance_policy_test.go (hunk 30 above)
Comment (barnuri-cp) [reply dome9/resource_dome9_image_assurance_policy_test.go]: fixed

File: dome9/resource_dome9_image_assurance_policy_test.go (hunk 27 above)
Comment (barnuri-cp) [reply dome9/resource_dome9_image_assurance_policy_test.go:53]: https://github.com/dome9/terraform-provider-dome9/pull/160/commits/4f1e668d2d46f3013836f2f3d92d5bb5f5a8e8a6

thats what you mean ?

File: dome9/resource_dome9_image_assurance_policy.go (hunk 25 above)
Comment (ronyscp) [reply dome9/resource_dome9_image_assurance_policy.go:76]: in AC the action is required.
what ` d.Get("admission_control_action")` returns? if it can be nil then `nil.(string)` will cause panic

**Comment by barnuri-cp** [comment]
run locally with @omartawfeeq  the unit tests and pass 
now testing locally with terraform

**Comment by barnuri-cp** [comment]
tested also with terraform locally on our prod env 

image

**Review by ronyscp — APPROVED** [review]
